clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java

```diff
@@ -146,6 +146,41 @@ R await(long timeout, TimeUnit unit)
         return future;
     }
 
+    private static class WhenCompleteBiConsumer<T> extends BiConsumer<T, Throwable> {
+        private final KafkaFutureImpl<T> future;
+        private final BiConsumer<T, Throwable> biConsumer;
+
+        WhenCompleteBiConsumer(KafkaFutureImpl<T> future, BiConsumer<T, Throwable> biConsumer) {
+            this.future = future;
+            this.biConsumer = biConsumer;
+        }
+
+        @Override
+        public void accept(T val, Throwable exception) {
+            try {
+                if (exception != null) {
+                    biConsumer.accept(null, exception);
+                } else {
+                    biConsumer.accept(val, null);
+                }
+            } catch (Throwable e) {
+                exception = e;
+            }
+            if (exception != null) {
+                future.completeExceptionally(exception);
+            } else {
+                future.complete(val);
+            }
+        }
+    }
+
+    @Override
+    public <T> KafkaFuture<T> whenComplete(final BiConsumer<? super T, ? super Throwable> biConsumer) {
+        final KafkaFutureImpl<T> future = new KafkaFutureImpl<T>();
+        addWaiter(new WhenCompleteBiConsumer(future, biConsumer));
```

Comment:
Compiler still gives some warning here, as you are erasing here the java generic type.  Adding the `<>`  infects the rest of your code with the `? super T` you have to add at the correct places.

Reply:
Fixed this in my latest update of #4033.